src/libsyntax/parse/token.rs

```diff
@@ -395,3 +452,47 @@ impl fmt::Debug for Nonterminal {
         }
     }
 }
+
+pub fn is_op(tok: &Token) -> bool {
+    match *tok {
+        OpenDelim(..) | CloseDelim(..) | Literal(..) | DocComment(..) |
+        Ident(..) | Underscore | Lifetime(..) | Interpolated(..) |
+        Whitespace | Comment | Shebang(..) | Eof => false,
+        _ => true,
+    }
+}
+
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct LazyTokenStream(RefCell<Option<TokenStream>>);
```

Comment:
As an aside: if the dynamic borrow checks end up too inefficient, then this could be re-implemented on top of [`MoveCell`](https://github.com/SimonSapin/rust-movecell).